{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_input_path <- function(sv_caller)\n",
    "{\n",
    "    paste(\n",
    "    '/Users/lsantuari/Documents/Manuscripts/sv-callers/Zenodo_data/data/benchmark/out/3/S3/',\n",
    "    sv_caller,'_out/',sv_caller,'.vcf', sep='')\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :\n",
      "“there is no package called ‘StructuralVariantAnnotation’”"
     ]
    }
   ],
   "source": [
    "suppressPackageStartupMessages(require(StructuralVariantAnnotation))\n",
    "\n",
    "#SV type assignment based on\n",
    "# https://github.com/PapenfussLab/gridss/blob/7b1fedfed32af9e03ed5c6863d368a821a4c699f/example/simple-event-annotation.R#L9\n",
    "apply_svtype <- function(gr)\n",
    "{\n",
    "  gr$svtype <-\n",
    "    ifelse(\n",
    "      seqnames(gr) != seqnames(partner(gr)),\n",
    "      \"BP\",\n",
    "      ifelse(\n",
    "        gr$insLen >= abs(gr$svLen) * 0.7,\n",
    "        \"INS\",\n",
    "        ifelse(\n",
    "          strand(gr) == strand(partner(gr)),\n",
    "          \"INV\",\n",
    "          ifelse(xor(\n",
    "            start(gr) < start(partner(gr)), strand(gr) == \"-\"\n",
    "          ), \"DEL\",\n",
    "          \"DUP\")\n",
    "        )\n",
    "      )\n",
    "    )\n",
    "  gr\n",
    "}\n",
    "\n",
    "load_sv_caller_vcf <-\n",
    "  function(vcf_file,\n",
    "           confidence_regions_gr,\n",
    "           sample,\n",
    "           sv_caller)\n",
    "  {\n",
    "    # vcf_file <- truth_set_file[[sample]]\n",
    "    sv_callset_vcf <-\n",
    "      VariantAnnotation::readVcf(vcf_file)\n",
    "    \n",
    "    if (sv_caller == 'lumpy')\n",
    "    {\n",
    "      # Read evidence support as a proxy for QUAL\n",
    "      support <- unlist(info(sv_callset_vcf)$SU)\n",
    "      fixed(sv_callset_vcf)$QUAL <- support\n",
    "    } else if (sv_caller == 'delly')\n",
    "    {\n",
    "      # Split-read support plus Paired-end read support as a proxy for QUAL\n",
    "      sr_support <- info(sv_callset_vcf)$SR\n",
    "      sr_support[is.na(sr_support)] <- 0\n",
    "      fixed(sv_callset_vcf)$QUAL <-\n",
    "        sr_support + info(sv_callset_vcf)$PE\n",
    "    }\n",
    "    \n",
    "    bpgr <- breakpointRanges(sv_callset_vcf)\n",
    "    begr <- breakendRanges(sv_callset_vcf)\n",
    "    gr <- sort(c(bpgr, begr))\n",
    "    if (sv_caller %in% c('gridss', 'manta')) {\n",
    "      gr <- apply_svtype(gr)\n",
    "    }\n",
    "    # Select DEL\n",
    "    gr <- gr[which(gr$svtype == \"DEL\")]\n",
    "    # Select DEL >= 50 bp\n",
    "    gr <- gr[gr$svLen <= (-50)]\n",
    "    gr\n",
    "  }\n",
    "\n",
    "load_truth_set_vcf <-\n",
    "  function(vcf_file, sample)\n",
    "  {\n",
    "    sv_callset_vcf <-\n",
    "      VariantAnnotation::readVcf(vcf_file)\n",
    "    bpgr <- breakpointRanges(sv_callset_vcf)\n",
    "    begr <- breakendRanges(sv_callset_vcf)\n",
    "    gr <- sort(c(bpgr, begr))\n",
    "\n",
    "    gr <- gr[which(gr$svtype == \"DEL\")]\n",
    "    gr <- gr[gr$svLen <= (-50)]\n",
    "    \n",
    "    gr <- remove_blacklist(gr, confidence_regions_gr, sample)\n",
    "    gr\n",
    "  }\n",
    "\n",
    "load_bedpe <- function(bedpe_file,\n",
    "                       sample)\n",
    "{\n",
    "  sv_callset_bedpe <- rtracklayer::import(bedpe_file)\n",
    "  bpgr <- pairs2breakpointgr(sv_callset_bedpe)\n",
    "  gr <- sort(bpgr)\n",
    "  gr <- remove_blacklist(gr, confidence_regions_gr, sample)\n",
    "  gr\n",
    "}\n",
    "\n",
    "make_percent <- function(x){\n",
    "  round(x*100,digits = 1)\n",
    "}\n",
    "\n",
    "datasets <- c('NA12878')\n",
    "\n",
    "truth_set_file <- list()\n",
    "truth_set_file[['NA12878']] <-\n",
    "  '/Users/lsantuari/Documents/Data/svclassify/Personalis_1000_Genomes_deduplicated_deletions.bedpe'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gr <- list()\n",
    "gr[[test_sample]] <- list()\n",
    "  \n",
    "truth_set <- list()\n",
    "  \n",
    "truth_set[[test_sample]] <-\n",
    "      load_bedpe(truth_set_file[[test_sample]], \n",
    "                 test_sample)\n",
    "\n",
    "#Specify output directory for results\n",
    "outDir <- \"\"\n",
    "\n",
    "sv_caller_list <- c('gridss', 'manta', 'lumpy', 'delly')\n",
    "\n",
    "# Load sv_callers results\n",
    "  for (sv_caller in sv_caller_list)\n",
    "  {\n",
    "    print(paste('Loading', sv_caller))\n",
    "\n",
    "    vcf_file <- get_input_path(sv_caller)\n",
    "      \n",
    "    print(paste('Loading ',vcf_file,sep=''))\n",
    "    gr[[test_sample]][[sv_caller]] <- load_sv_caller_vcf(vcf_file, test_sample, sv_caller)\n",
    "  }\n",
    "\n",
    "  # Only consider Chr1 to ChrX\n",
    "  for (sv_caller in sv_caller_list)\n",
    "  {\n",
    "    print(sv_caller)\n",
    "    print(length(gr[[test_sample]][[sv_caller]]))\n",
    "    print(length(gr[[test_sample]][[sv_caller]][seqnames(gr[[test_sample]][[sv_caller]]) !=\n",
    "                                             'Y']))\n",
    "  }\n",
    "\n",
    "  truth_svgr <- truth_set[[test_sample]]\n",
    "  for (sv_caller in sv_caller_list)\n",
    "  {\n",
    "    gr[[test_sample]][[sv_caller]]$caller <- sv_caller\n",
    "  }\n",
    "\n",
    "  for (svcaller in sv_caller_list)\n",
    "  {\n",
    "    gr[[test_sample]][[svcaller]]$truth_matches <-\n",
    "      countBreakpointOverlaps(\n",
    "        gr[[test_sample]][[svcaller]],\n",
    "        truth_svgr,\n",
    "        # read pair based callers make imprecise calls.\n",
    "        # A margin around the call position is required when matching with the truth set\n",
    "        maxgap = 200,\n",
    "        # Since we added a maxgap, we also need to restrict the mismatch between the\n",
    "        # size of the events. We don't want to match a 100bp deletion with a\n",
    "        # 5bp duplication. This will happen if we have a 100bp margin but don't also\n",
    "        # require an approximate size match as well\n",
    "        sizemargin = 0.25,\n",
    "        ignore.strand = TRUE,\n",
    "        # We also don't want to match a 20bp deletion with a 20bp deletion 80bp away\n",
    "        # by restricting the margin based on the size of the event, we can make sure\n",
    "        # that simple events actually do overlap\n",
    "        restrictMarginToSizeMultiple =\n",
    "          0.5,\n",
    "        # Some callers make duplicate calls and will sometimes report a variant multiple\n",
    "        # times with slightly different bounds. countOnlyBest prevents these being\n",
    "        # double-counted as multiple true positives.\n",
    "        countOnlyBest = TRUE\n",
    "      )\n",
    "  }\n",
    "\n",
    "    svgr <- c(gr[[test_sample]][['gridss']],\n",
    "              gr[[test_sample]][['manta']],\n",
    "              gr[[test_sample]][['lumpy']],\n",
    "              gr[[test_sample]][['delly']])\n",
    "\n",
    "  res.df <- as.data.frame(svgr) %>%\n",
    "    dplyr::select(caller, truth_matches) %>%\n",
    "    dplyr::group_by(caller) %>%\n",
    "    dplyr::summarise(calls = n(),\n",
    "                     tp = sum(truth_matches > 0)) %>%\n",
    "    dplyr::group_by(caller) %>%\n",
    "    dplyr::mutate(\n",
    "      cum_tp = cumsum(tp),\n",
    "      cum_n = cumsum(calls),\n",
    "      cum_fp = cum_n - cum_tp,\n",
    "      precision = round(cum_tp / cum_n,digits = 1),\n",
    "      recall = round(cum_tp / length(truth_svgr), digits = 1)\n",
    "    )\n",
    "  res.df$F1 = with(res.df, 2 * (precision * recall) / (precision + recall))\n",
    "\n",
    "  res.df$precision <- make_percent(res.df$precision)\n",
    "  res.df$recall <- make_percent(res.df$recall)\n",
    "  res.df$F1 <- make_percent(res.df$F1)\n",
    "\n",
    "  write.table(res.df, file=paste(outDir, 'performance_results.csv',sep=''), quote=F, row.names = F)\n",
    "\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
